### PR TITLE
Introducing Windows Customization

### PR DESCRIPTION
**Pull request by nieubank**
## Summary of the pull request
This commit includes the initial placeholder L1 page for 'Windows customization' which will host various settings and utilities to customize Windows to a developer's preferences.

image

image

## References and relevant issues
Starts the work related to the "Advanced Settings" feature exploration #1983 

## Detailed description of the pull request / Additional comments
The MainPage acts as a root which eventually will host a ContentPresenter for displaying the active view (for example, when search is added the ContentPresent will switch the root view temporarily to the 'All settings' search view, and once we get better support to integrate into the navigation service, sub-pages could be consolidated into that content presenter as well.

To provide an example of how options can be incorporated, the Developer File Explorer page was introduced.

The pattern we’re going for here is to provide logical settings groups in the form of a UserControl, which is then wrapped by a Page. This helps in the future when implementing search because we want to host all settings controls in the same root view and filter visibility based on the search filter. Having each group as a separate control view can easily enable hosting that UI in any form we need to down the line. The sub-pages themselves are probably temporarily placeholders for a better integration with the navigation service down the line.

See tools\Customization\DevHome.Customization\Customization.md for details.

## Validation steps performed
- Verified that the Windows customization section appears on the navigation menu and can be navigated to.
- Clicking on the Windows developer settings link will navigate to the Windows Settings app "For developers" page
- Toggle each setting on "Developer File Explorer" and verify that the setting takes effect, both on and off
- 
## PR checklist
- [x] Makes progress on #1983 
- [x] Tests passed
- [x] Documentation updated

### CODE REVIEW

**Comment by nieubank** [comment]
Waiting on the new DevHomeHelpers nuget to become available.

File: src/Services/PageService.cs
Comment (krschau) [inline src/Services/PageService.cs:43]: This can change after https://github.com/microsoft/devhome/pull/2356 goes in, hopefully later this week.

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs
Comment (krschau) [inline tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs:32]: Just as an FYI, if you wanted this stuff in the ViewModel, you can do the same thing we did in #2313 to move it over and use the built-in headers. But you might want to hold off on that, since we're considering changing the app-wide header template to be a BreadcrumbBar (input welcome).

File: tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs
Comment (krschau) [inline tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs:15]: Consider renaming MainPage* to CustomizationMainPage* (or something). There's also a MainPageViewModel, etc in SetupFlow, and might be useful to differentiate.

File: src/Styles/TextBlock.xaml
Comment (nieubank) [reply src/Styles/TextBlock.xaml:41]: It is a bit strange, but this follows the pattern in SettingsSectionHeaderTextBlockStyle. I'm not sure historically why that one has an offset of 1, but if that needs to stay, this looks odd without it.

File: tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs
Comment (nieubank) [reply tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs:15]: I did consider that but then realized that the name-spacing makes that redundant so unless you have a strong preference, I think this is more concise.

**Review by skanda890 — COMMENTED** [review]
Should implement this.

File: tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs:38]: Why make a static method to log this event?  `TelemetryFactory` can be accessed from outside this class, ya?

File: tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/ViewModels/MainPageViewModel.cs:30]: Why not use `IHost` in the constructor to get the `INavigationService` and store it in a `readonly ` variable?

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml:12]: Why are you setting `MaxWidth`

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs:21]: I would nudge you to move `Breadcrumbs` to the `ViewModel`

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs:43]: c-style casts can throw.  Maybe use `as` then check for `is not null`?

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerView.xaml.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerView.xaml.cs:18]: Suggest using `IHost` here to get the `ViewModel`

`Application.Current` throws an exception when ran in a unit test.

File: tools/Customization/DevHome.Customization/Views/MainPage.cs
Comment (dhoehna) [inline tools/Customization/DevHome.Customization/Views/MainPage.cs:17]: Does this string need to be localized?

Comment (nieubank) [reply tools/Customization/DevHome.Customization/Views/MainPage.cs:17]: Good question @krschau / @EricJohnson327: none of these are localized across the project, what's the usage for ShortName?

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml
Comment (nieubank) [reply tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml:12]: This seems to be standard practice with ToolPages, just following preexisting patterns.

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs
Comment (krschau) [reply tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs:21]: @dhoehna https://github.com/microsoft/devhome/pull/2383#discussion_r1520332745

File: tools/Customization/DevHome.Customization/Views/MainPage.cs
Comment (krschau) [reply tools/Customization/DevHome.Customization/Views/MainPage.cs:17]: I was just asking @EricJohnson327 this on Friday. We think it's left over from an earlier version of navigation. It doesn't have to be localized because it's not used anymore -- instead, the text in the left nav comes from the resw file. At some point we should remove ShortName altogether.

File: src/Services/PageService.cs
Comment (krschau) [reply src/Services/PageService.cs:43]: #2356 is checked in now.

File: tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs
Comment (nieubank) [reply tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs:38]: Yes, but I want to make sure the log metadata is consistent so this helps accomplish that as a wrapper.

File: tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs
Comment (nieubank) [inline tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs:40]: >Critical [](http://example.com/codeflow?start=94&length=8)

should probably be a measure

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerView.xaml.cs
Comment (nieubank) [reply tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerView.xaml.cs:18]: Is there an example of a View doing this, they need a default constructor so I'm not sure how to pull that off offhand. I would suggest we file a follow-up task to do this across the project since most if not all existing ViewModels are acquired with this pattern.

File: tools/Customization/DevHome.Customization/DevHome.Customization.csproj
Comment (krschau) [inline tools/Customization/DevHome.Customization/DevHome.Customization.csproj:21]: If you remove the (unused) references to DevHome.Settings.Models, you can remove this ProjectReference. There seem to be a bunch of unused usings left over, but this is the only one that really matters.

File: tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs
Comment (krschau) [inline tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs:40]: At this time I'm not sure any of these will come through, we've gotten approvals to use LogLevel.Critical around dev home. You do have to get approvals for each event, though.

**Review by krschau — APPROVED** [review]
Approving but strongly suggest making the ProjectReference change.

File: tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs
Comment (nieubank) [reply tools/Customization/DevHome.Customization/Views/DeveloperFileExplorerPage.xaml.cs]: Decided to just pull the Breadcrumbs for now, we can revisit a more generic solution that would apply here after this goes in.

File: tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs
Comment (nieubank) [reply tools/Customization/DevHome.Customization/TelemetryEvents/SettingChangedEvent.cs:40]: What is the recommended approach here? We will want to set up measures for these at some point but I'm not sure it's critical to do that until we have a good telemetry story for Windows customization. I could just change this to be Info for now.

**Review by EricJohnson327 — APPROVED** [review]

**Pull request merged**
